contracts/integrations/test/framework-unit-tests/function_assertion_test.ts

```diff
@@ -0,0 +1,125 @@
+                    sideEffectTarget = beforeInfo;
+                },
+            });
+            await assertion.runAsync(randomInput);
+            expect(sideEffectTarget).bignumber.to.be.eq(randomInput);
+        });
+
+        it('should pass the result from the function call to "after"', async () => {
+            let sideEffectTarget = constants.ZERO_AMOUNT;
+            const assertion = new FunctionAssertion(exampleContract.revertSideEffect, {
+                before: async (inputValue: BigNumber) => {},
+                after: async (beforeInfo: any, result: Result, returnValue: BigNumber) => {
+                    sideEffectTarget = result.data;
+                },
+            });
+            const randomInput = generatePseudoRandom256BitNumber();
+            await assertion.runAsync(randomInput);
+            expect(sideEffectTarget).bignumber.to.be.eq(randomInput);
+        });
+
+        it('should pass the receipt from the function call to "after"', async () => {
+            let sideEffectTarget: TransactionReceiptWithDecodedLogs = {} as TransactionReceiptWithDecodedLogs;
+            const assertion = new FunctionAssertion(exampleContract.revertSideEffect, {
+                before: async (inputValue: BigNumber) => {},
+                after: async (beforeInfo: any, result: Result, returnValue: BigNumber) => {
+                    if (result.receipt) {
+                        sideEffectTarget = result.receipt;
+                    } else {
+                        throw new Error('No receipt received.');
+                    }
+                },
+            });
+            const randomInput = generatePseudoRandom256BitNumber();
+            await assertion.runAsync(randomInput);
+
+            // Ensure that the correct events were emitted.
+            const [event] = filterLogsToArguments<TestFrameworkSomeEventEventArgs>(
+                sideEffectTarget.logs,
+                TestFrameworkEvents.SomeEvent,
+            );
+            expect(event).to.be.deep.eq({ someNumber: randomInput });
+        });
+
+        it('should pass the error to "after" if the function call fails', async () => {
+            let sideEffectTarget: Error = new Error('');
+            await exampleContract.setCounter.awaitTransactionSuccessAsync(new BigNumber(1));
+            const assertion = new FunctionAssertion(exampleContract.revertSideEffect, {
+                before: async (inputValue: BigNumber) => {},
+                after: async (beforeInfo: any, result: Result, returnValue: BigNumber) => {
+                    sideEffectTarget = result.data;
+                },
+            });
+            const randomInput = generatePseudoRandom256BitNumber();
+            await assertion.runAsync(randomInput);
+            const errorMessage = 'VM Exception while processing transaction: revert Revert';
+            expect(sideEffectTarget.message).to.be.eq(errorMessage);
```

Comment:
`result.data` in this case should be a `RevertError` type, right? I wonder if we can just do `expect(sideEffectTarget).to.revertWith('Revert')` here.

Reply:
I wasn't able to get that to work because `sideEffectTarget` is not a promise. I'll mess with it a bit, and see if I can figure something out.